sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveSharedStateSuite.scala

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hive
+
+import org.apache.hadoop.hive.conf.HiveConf.ConfVars
+
+import org.apache.spark.{SparkConf, SparkContext, SparkFunSuite}
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.internal.SharedState
+import org.apache.spark.sql.internal.StaticSQLConf._
+import org.apache.spark.util.Utils
+
+class HiveSharedStateSuite extends SparkFunSuite {
+
+  test("enableHiveSupport has right to determine the catalog while using an existing sc") {
+    val conf = new SparkConf().setMaster("local").setAppName("SharedState Test")
+    val sc = SparkContext.getOrCreate(conf)
+    val ss = SparkSession.builder().enableHiveSupport().getOrCreate()
+    assert(ss.sharedState.externalCatalog.unwrapped.isInstanceOf[HiveExternalCatalog],
+      "The catalog should be hive ")
+
+    val ss2 = SparkSession.builder().getOrCreate()
+    assert(ss2.sharedState.externalCatalog.unwrapped.isInstanceOf[HiveExternalCatalog],
+      "The catalog should be shared across sessions")
+  }
+
+  test("initial configs should be passed to SharedState but not SparkContext") {
+    val conf = new SparkConf().setMaster("local").setAppName("SharedState Test")
+    val sc = SparkContext.getOrCreate(conf)
+    val invalidPath = "invalid/path"
+    val metastorePath = Utils.createTempDir()
+    val tmpDb = "tmp_db"
+
+    // The initial configs used to generate SharedState, none of these should affect the global
+    // shared SparkContext's configurations. Especially, all these configs are passed to the cloned
+    // confs inside SharedState except metastore warehouse dir.
+    val initialConfigs = Map("spark.foo" -> "bar",
+      WAREHOUSE_PATH.key -> invalidPath,
+      ConfVars.METASTOREWAREHOUSE.varname -> invalidPath,
+      CATALOG_IMPLEMENTATION.key -> "hive",
+      ConfVars.METASTORECONNECTURLKEY.varname ->
+        s"jdbc:derby:;databaseName=$metastorePath/metastore_db;create=true",
+      GLOBAL_TEMP_DATABASE.key -> tmpDb)
+
+    val state = new SharedState(sc, initialConfigs)
```

Comment:
Can we be more real-world and create SparkSession here?

Reply:
In hive module test, there is an long cached TestHiveSession, so I use this SharedState explicitly

Reply:
why can we build SparkSession in the above test case?

Reply:
Individually the above test runs correctly, but will use a existing SparkSession with hive support while runs under the whole hive module. so I add the second test to insure  all configurations passed correctly.

Reply:
shall we add `beforeEach` and call `SparkSession.clearActiveSession` and `SparkSession.clearDefaultSession`?

Reply:
if we clear session here then dozens of existing tests will fail. Or we cached that active session here and set it back in `afterEach` by calling `SparkSession.setActiveSession`?

Reply:
Em.. It can't be done since `spark.driver.allowMultiContext` no longer exists any more.

Reply:
then shall we move the test to sql core module? We can add try-catch and make sure ClassNotFound exception is thrown.

Reply:
yes, we can move the first test to sql core module, and may leave the second in hive module to ensure all confs correct.

Reply:
I tried to move test 1 to sql module but it will fail fast with `IllegalArgumentException` in enableHiveSupport without hive class present, classnotfound can't be captured.

Reply:
can we change test 1 to create `SharedState` as well?

Reply:
the intent of test 1 is used to verify enableHiveSupport. I think `SharedState` is already covered in test 2. Or maybe I can just delete test 1 ?

Reply:
test 1 can pass without your fix, right?

Reply:
without this fix ，test 1 fails individually but pass fully

Reply:
I'd like to move test 1 to sql core module, we can catch the `IllegalArgumentException` to check it. Then the test is reliable, it doesn't matter we run it individually or with other tests.

Reply:
IllegalArgumentException does not mean that the catalog is hive or not. It will also pass anyway with or without this fix，and the fact is that it doesn't have the chance to get to where the catalog is initialized.

Reply:
then let's just remove test 1, but post the code in PR description to let people know how to reproduce the bug.

Reply:
ok. both done.